File: exercise/php/day13/tests/SantaMarketTest.php
<?php

use SantaMarket\ChristmasElf;
use SantaMarket\Discount;
use SantaMarket\Product;
use SantaMarket\ProductUnit;
use SantaMarket\ReceiptItem;
use SantaMarket\ShoppingSleigh;
use SantaMarket\SpecialOfferType;
use Tests\FakeCatalog;

function expectFloatCloseTo($actual, $expected, $tolerance = 0.001): void
{
    expect(abs($actual - $expected))->toBeLessThanOrEqual($tolerance);
}

function expectDiscountCloseTo(Discount $actual, Discount $expected, $tolerance = 0.001): void
{
    expect($actual->getDescription())->toEqual($expected->getDescription())
        ->and(abs($actual->getDiscountAmount() - $expected->getDiscountAmount()))->toBeLessThanOrEqual($tolerance)
        ->and($actual->getProduct())->toEqual($expected->getProduct());
}

it('calculates total without discounts', function () {
    $catalog = new FakeCatalog();
    $teddyBear = new Product('teddyBear', ProductUnit::EACH);
    $teddyBearPrice = 1.0;
    $catalog->addProduct($teddyBear, $teddyBearPrice);

    $sleigh = new ShoppingSleigh();
    $numberOfTeddyBears = 3;
    $sleigh->addItemQuantity($teddyBear, $numberOfTeddyBears);

    $elf = new ChristmasElf($catalog);
    $receipt = $elf->checksOutArticlesFrom($sleigh);

    $expectedTotalPrice = $numberOfTeddyBears * $teddyBearPrice;
    $expectedReceiptItem = new ReceiptItem($teddyBear, $numberOfTeddyBears, $teddyBearPrice, $expectedTotalPrice);

    expectFloatCloseTo($receipt->getTotalPrice(), $expectedTotalPrice);
    expect($receipt->getItems()[0])->toEqual($expectedReceiptItem);
});

it('applies ten percent discount', function () {
    $catalog = new FakeCatalog();
    $turkey = new Product('turkey', ProductUnit::KILO);
    $turkeyPrice = 2.0;
    $catalog->addProduct($turkey, $turkeyPrice);

    $sleigh = new ShoppingSleigh();
    $turkeyQuantity = 2.0;
    $sleigh->addItemQuantity($turkey, $turkeyQuantity);

    $elf = new ChristmasElf($catalog);
    $elf->addSpecialOffer(SpecialOfferType::TEN_PERCENT_DISCOUNT, $turkey, 10.0);
    $receipt = $elf->checksOutArticlesFrom($sleigh);

    $expectedNonDiscountedPrice = $turkeyQuantity * $turkeyPrice;
    $expectedTotalPrice = $expectedNonDiscountedPrice * 0.9;
    $expectedReceiptItem = new ReceiptItem($turkey, $turkeyQuantity, $turkeyPrice, $expectedNonDiscountedPrice);
    $expectedDiscount = new Discount($turkey, '10% off', round($expectedTotalPrice - $expectedNonDiscountedPrice, 2));

    expectFloatCloseTo($receipt->getTotalPrice(), $expectedTotalPrice);
    expectDiscountCloseTo($receipt->getDiscounts()[0], $expectedDiscount);
    expect($receipt->getItems()[0])->toEqual($expectedReceiptItem);
});

it('applies three for two discount', function () {
    $catalog = new FakeCatalog();
    $teddyBear = new Product('teddyBear', ProductUnit::EACH);
    $teddyBearPrice = 1.0;
    $catalog->addProduct($teddyBear, $teddyBearPrice);

    $elf = new ChristmasElf($catalog);
    $elf->addSpecialOffer(SpecialOfferType::THREE_FOR_TWO, $teddyBear, 0.0);

    $sleigh = new ShoppingSleigh();
    $numberOfTeddyBears = 3;
    $sleigh->addItemQuantity($teddyBear, $numberOfTeddyBears);

    $receipt = $elf->checksOutArticlesFrom($sleigh);

    $expectedNonDiscountedPrice = $numberOfTeddyBears * $teddyBearPrice;
    $expectedTotalPrice = 2 * $teddyBearPrice;
    $expectedReceiptItem = new ReceiptItem($teddyBear, $numberOfTeddyBears, $teddyBearPrice, $expectedNonDiscountedPrice);
    $expectedDiscount = new Discount($teddyBear, '3 for 2', $expectedTotalPrice - $expectedNonDiscountedPrice);

    expectFloatCloseTo($receipt->getTotalPrice(), $expectedTotalPrice);
    expectDiscountCloseTo($receipt->getDiscounts()[0], $expectedDiscount);
    expect($receipt->getItems()[0])->toEqual($expectedReceiptItem);
});

it('applies two for amount discount', function () {
    $catalog = new FakeCatalog();
    $teddyBear = new Product('teddyBear', ProductUnit::EACH);
    $teddyBearPrice = 1.0;
    $catalog->addProduct($teddyBear, $teddyBearPrice);

    $elf = new ChristmasElf($catalog);
    $discountedPriceForTwoTeddyBears = 1.6;
    $elf->addSpecialOffer(SpecialOfferType::TWO_FOR_AMOUNT, $teddyBear, $discountedPriceForTwoTeddyBears);

    $sleigh = new ShoppingSleigh();
    $numberOfTeddyBears = 2;
    $sleigh->addItemQuantity($teddyBear, $numberOfTeddyBears);

    $receipt = $elf->checksOutArticlesFrom($sleigh);

    $expectedNonDiscountedPrice = $numberOfTeddyBears * $teddyBearPrice;
    $expectedTotalPrice = $discountedPriceForTwoTeddyBears;
    $expectedReceiptItem = new ReceiptItem($teddyBear, $numberOfTeddyBears, $teddyBearPrice, $expectedNonDiscountedPrice);
    $expectedDiscount = new Discount($teddyBear, '2 for ' . $discountedPriceForTwoTeddyBears, $expectedTotalPrice - $expectedNonDiscountedPrice);

    expectFloatCloseTo($receipt->getTotalPrice(), $expectedTotalPrice);
    expectDiscountCloseTo($receipt->getDiscounts()[0], $expectedDiscount);
    expect($receipt->getItems()[0])->toEqual($expectedReceiptItem);
});

it('applies five for amount discount', function () {
    $catalog = new FakeCatalog();
    $teddyBear = new Product('teddyBear', ProductUnit::EACH);
    $teddyBearPrice = 1.0;
    $catalog->addProduct($teddyBear, $teddyBearPrice);

    $elf = new ChristmasElf($catalog);
    $discountedPriceForFiveTeddyBears = 4.0;
    $elf->addSpecialOffer(SpecialOfferType::FIVE_FOR_AMOUNT, $teddyBear, $discountedPriceForFiveTeddyBears);

    $sleigh = new ShoppingSleigh();
    $numberOfTeddyBears = 6;
    for ($i = 0; $i < $numberOfTeddyBears; $i++) {
        $sleigh->addItem($teddyBear);
    }

    $receipt = $elf->checksOutArticlesFrom($sleigh);

    $expectedNonDiscountedPrice = $numberOfTeddyBears * $teddyBearPrice;
    $expectedTotalPrice = $discountedPriceForFiveTeddyBears + $teddyBearPrice;
    $expectedReceiptItem = new ReceiptItem($teddyBear, 1, $teddyBearPrice, $teddyBearPrice);
    $expectedDiscount = new Discount($teddyBear, '5 for ' . $discountedPriceForFiveTeddyBears, $expectedTotalPrice - $expectedNonDiscountedPrice);

    expectFloatCloseTo($receipt->getTotalPrice(), $expectedTotalPrice);
    expectDiscountCloseTo($receipt->getDiscounts()[0], $expectedDiscount);
    expect($receipt->getItems()[0])->toEqual($expectedReceiptItem);
});

it('applies discount for one of two different items in sleigh', function () {
    $catalog = new FakeCatalog();
    $teddyBear = new Product('teddyBear', ProductUnit::EACH);
    $teddyBearPrice = 1.0;
    $catalog->addProduct($teddyBear, $teddyBearPrice);

    $turkey = new Product('turkey', ProductUnit::KILO);
    $turkeyPrice = 2.0;
    $catalog->addProduct($turkey, $turkeyPrice);

    $elf = new ChristmasElf($catalog);
    $elf->addSpecialOffer(SpecialOfferType::TEN_PERCENT_DISCOUNT, $teddyBear, 10.0);

    $sleigh = new ShoppingSleigh();
    $sleigh->addItemQuantity($teddyBear, 3);
    $sleigh->addItemQuantity($turkey, 1.5);

    $receipt = $elf->checksOutArticlesFrom($sleigh);

    $expectedNonDiscountedTeddyBearPrice = 3 * $teddyBearPrice;
    $expectedTotalTeddyBearPrice = $expectedNonDiscountedTeddyBearPrice * 0.9;
    $totalTurkeyPrice = $turkeyPrice * 1.5;
    $expectedTotalPrice = $expectedTotalTeddyBearPrice + $totalTurkeyPrice;

    $expectedDiscount = new Discount($teddyBear, '10% off', $expectedTotalTeddyBearPrice - $expectedNonDiscountedTeddyBearPrice);

    expectFloatCloseTo($receipt->getTotalPrice(), $expectedTotalPrice);
    expectDiscountCloseTo($receipt->getDiscounts()[0], $expectedDiscount);
});

File: solution/day13/src/test/java/santamarket/model/SantamarketTest.java
package santamarket.model;

import org.junit.jupiter.api.Test;

import java.util.Collections;
import java.util.List;

import static org.hamcrest.MatcherAssert.assertThat;
import static santamarket.model.ReceiptMatcher.matches;

class SantamarketTest {
    @Test
    void noDiscount() {
        SantamarketCatalog catalog = new FakeCatalog();

        Product teddyBear = new Product("teddyBear", ProductUnit.EACH);
        double teddyBearPrice = 1;
        catalog.addProduct(teddyBear, teddyBearPrice);

        ShoppingSleigh sleigh = new ShoppingSleigh();
        int numberOfTeddyBears = 3;
        sleigh.addItemQuantity(teddyBear, numberOfTeddyBears);

        // ACT
        ChristmasElf elf = new ChristmasElf(catalog);
        Receipt receipt = elf.checksOutArticlesFrom(sleigh);

        // ASSERT
        double expectedTotalPrice = numberOfTeddyBears * teddyBearPrice;
        ReceiptItem expectedReceiptItem = new ReceiptItem(teddyBear, numberOfTeddyBears, teddyBearPrice, expectedTotalPrice);

        ReceiptMatcher matcher = matches(expectedTotalPrice, List.of(expectedReceiptItem));
        assertThat(receipt, matcher);
    }

    @Test
    void tenPercentDiscount() {
        SantamarketCatalog catalog = new FakeCatalog();

        Product turkey = new Product("turkey", ProductUnit.KILO);
        double turkeyPrice = 2.;
        catalog.addProduct(turkey, turkeyPrice);

        ShoppingSleigh sleigh = new ShoppingSleigh();
        double turkeyQuantity = 2;
        sleigh.addItemQuantity(turkey, turkeyQuantity);

        ChristmasElf elf = new ChristmasElf(catalog);
        elf.addSpecialOffer(SpecialOfferType.TEN_PERCENT_DISCOUNT, turkey, 10.0);

        // ACT
        Receipt receipt = elf.checksOutArticlesFrom(sleigh);

        // ASSERT
        double expectedNonDiscountedPrice = turkeyQuantity * turkeyPrice;
        double expectedTotalPrice = expectedNonDiscountedPrice * 0.9;
        ReceiptItem expectedReceiptItem = new ReceiptItem(turkey, turkeyQuantity, turkeyPrice, expectedNonDiscountedPrice);
        Discount expectedDiscount = new Discount(turkey, "10.0% off", expectedTotalPrice - expectedNonDiscountedPrice);
        ReceiptMatcher matcher = matches(expectedTotalPrice, List.of(expectedReceiptItem), List.of(expectedDiscount));
        assertThat(receipt, matcher);
    }

    @Test
    void threeForTwoDiscount() {
        SantamarketCatalog catalog = new FakeCatalog();

        Product teddyBear = new Product("teddyBear", ProductUnit.EACH);
        double teddyBearPrice = 1;
        catalog.addProduct(teddyBear, teddyBearPrice);

        ChristmasElf elf = new ChristmasElf(catalog);
        elf.addSpecialOffer(SpecialOfferType.THREE_FOR_TWO, teddyBear, 0.);

        ShoppingSleigh sleigh = new ShoppingSleigh();
        int numberOfTeddyBears = 3;
        sleigh.addItemQuantity(teddyBear, numberOfTeddyBears);

        // ACT
        Receipt receipt = elf.checksOutArticlesFrom(sleigh);

        // ASSERT
        double expectedNonDiscountedPrice = numberOfTeddyBears * teddyBearPrice;
        double expectedTotalPrice = 2 * teddyBearPrice;
        ReceiptItem expectedReceiptItem = new ReceiptItem(teddyBear, numberOfTeddyBears, teddyBearPrice, expectedNonDiscountedPrice);
        Discount expectedDiscount = new Discount(teddyBear, "3 for 2", expectedTotalPrice - expectedNonDiscountedPrice);
        ReceiptMatcher matcher = matches(expectedTotalPrice, List.of(expectedReceiptItem), List.of(expectedDiscount));
        assertThat(receipt, matcher);
    }

    @Test
    void twoForAmountDiscount() {
        SantamarketCatalog catalog = new FakeCatalog();

        Product teddyBear = new Product("teddyBear", ProductUnit.EACH);
        double teddyBearPrice = 1;
        catalog.addProduct(teddyBear, teddyBearPrice);

        ChristmasElf elf = new ChristmasElf(catalog);
        double discountedPriceForTwoTeddyBears = 1.6;
        elf.addSpecialOffer(SpecialOfferType.TWO_FOR_AMOUNT, teddyBear, discountedPriceForTwoTeddyBears);

        ShoppingSleigh sleigh = new ShoppingSleigh();
        int numberOfTeddyBears = 2;
        sleigh.addItemQuantity(teddyBear, numberOfTeddyBears);

        // ACT
        Receipt receipt = elf.checksOutArticlesFrom(sleigh);

        // ASSERT
        double expectedNonDiscountedPrice = numberOfTeddyBears * teddyBearPrice;
        double expectedTotalPrice = discountedPriceForTwoTeddyBears;
        ReceiptItem expectedReceiptItem = new ReceiptItem(teddyBear, numberOfTeddyBears, teddyBearPrice, expectedNonDiscountedPrice);
        Discount expectedDiscount = new Discount(teddyBear, "2 for " + discountedPriceForTwoTeddyBears, expectedTotalPrice - expectedNonDiscountedPrice);
        ReceiptMatcher matcher = matches(expectedTotalPrice, List.of(expectedReceiptItem), List.of(expectedDiscount));
        assertThat(receipt, matcher);
    }

    @Test
    void fiveForAmountDiscount() {
        SantamarketCatalog catalog = new FakeCatalog();

        Product teddyBear = new Product("teddyBear", ProductUnit.EACH);
        double teddyBearPrice = 1;
        catalog.addProduct(teddyBear, teddyBearPrice);

        ChristmasElf elf = new ChristmasElf(catalog);
        double discountedPriceForFiveTeddyBears = 4;
        elf.addSpecialOffer(SpecialOfferType.FIVE_FOR_AMOUNT, teddyBear, discountedPriceForFiveTeddyBears);

        ShoppingSleigh sleigh = new ShoppingSleigh();
        int numberOfTeddyBears = 6;
        for (int i = 0; i < numberOfTeddyBears; i++) {
            sleigh.addItem(teddyBear);
        }

        // ACT
        Receipt receipt = elf.checksOutArticlesFrom(sleigh);

        // ASSERT
        double expectedNonDiscountedPrice = numberOfTeddyBears * teddyBearPrice;
        double expectedTotalPrice = discountedPriceForFiveTeddyBears + teddyBearPrice;
        ReceiptItem expectedReceiptItem = new ReceiptItem(teddyBear, 1, teddyBearPrice, teddyBearPrice);
        Discount expectedDiscount = new Discount(teddyBear, "5 for " + discountedPriceForFiveTeddyBears, expectedTotalPrice - expectedNonDiscountedPrice);
        ReceiptMatcher matcher = matches(expectedTotalPrice, Collections.nCopies(numberOfTeddyBears, expectedReceiptItem), List.of(expectedDiscount));
        assertThat(receipt, matcher);
    }

    @Test
    void twoDifferentItemsInSleighWithDiscountOnOneOfThem() {
        SantamarketCatalog catalog = new FakeCatalog();

        Product teddyBear = new Product("teddyBear", ProductUnit.EACH);
        double teddyBearPrice = 1;
        catalog.addProduct(teddyBear, teddyBearPrice);

        Product turkey = new Product("turkey", ProductUnit.KILO);
        double turkeyPrice = 2.;
        catalog.addProduct(turkey, turkeyPrice);

        ChristmasElf christmasElf = new ChristmasElf(catalog);
        christmasElf.addSpecialOffer(SpecialOfferType.TEN_PERCENT_DISCOUNT, teddyBear, 10.0);

        ShoppingSleigh sleigh = new ShoppingSleigh();
        int numberOfTeddyBears = 3;
        sleigh.addItemQuantity(teddyBear, numberOfTeddyBears);
        double turkeyQuantity = 1.5;
        sleigh.addItemQuantity(turkey, turkeyQuantity);

        // ACT
        Receipt receipt = christmasElf.checksOutArticlesFrom(sleigh);

        // ASSERT
        double expectedNonDiscountedTeddyBearPrice = numberOfTeddyBears * teddyBearPrice;
        ReceiptItem teddyBearReceiptItem = new ReceiptItem(teddyBear, numberOfTeddyBears, teddyBearPrice, expectedNonDiscountedTeddyBearPrice);
        double totalTurkeyPrice = turkeyPrice * turkeyQuantity;
        ReceiptItem turkeyReceiptItem = new ReceiptItem(turkey, turkeyQuantity, turkeyPrice, totalTurkeyPrice);
        double expectedTotalTeddyBearPrice = expectedNonDiscountedTeddyBearPrice * 0.9;
        double expectedTotalPrice = expectedTotalTeddyBearPrice + totalTurkeyPrice;
        Discount expectedDiscount = new Discount(teddyBear, "10.0% off", expectedTotalTeddyBearPrice - expectedNonDiscountedTeddyBearPrice);
        ReceiptMatcher matcher = matches(expectedTotalPrice, List.of(teddyBearReceiptItem, turkeyReceiptItem), List.of(expectedDiscount));
        assertThat(receipt, matcher);
    }

    @Test
    void twoForOne() {
        SantamarketCatalog catalog = new FakeCatalog();

        Product teddyBear = new Product("teddyBear", ProductUnit.EACH);
        double teddyBearPrice = 1;
        catalog.addProduct(teddyBear, teddyBearPrice);

        ChristmasElf christmasElf = new ChristmasElf(catalog);
        christmasElf.addSpecialOffer(SpecialOfferType.TWO_FOR_ONE, teddyBear, 0.);

        ShoppingSleigh sleigh = new ShoppingSleigh();
        int numberOfTeddyBears = 4;
        sleigh.addItemQuantity(teddyBear, numberOfTeddyBears);

        // ACT
        Receipt receipt = christmasElf.checksOutArticlesFrom(sleigh);

        // ASSERT
        double expectedNonDiscountedPrice = numberOfTeddyBears * teddyBearPrice;
        double expectedTotalPrice = 2 * teddyBearPrice;
        ReceiptItem expectedReceiptItem = new ReceiptItem(teddyBear, numberOfTeddyBears, teddyBearPrice, expectedNonDiscountedPrice);
        Discount expectedDiscount = new Discount(teddyBear, "2 for 1", expectedTotalPrice - expectedNonDiscountedPrice);
        ReceiptMatcher matcher = matches(expectedTotalPrice, List.of(expectedReceiptItem), List.of(expectedDiscount));
        assertThat(receipt, matcher);
    }
}